File: database/database-annotation-processor/src/main/java/ru/tinkoff/kora/database/annotation/processor/cassandra/UserDefinedTypeStatementSetterGenerator.java
package ru.tinkoff.kora.database.annotation.processor.cassandra;


import com.squareup.javapoet.*;
import ru.tinkoff.kora.annotation.processor.common.CommonClassNames;
import ru.tinkoff.kora.annotation.processor.common.CommonUtils;
import ru.tinkoff.kora.database.annotation.processor.entity.DbEntity;

import javax.annotation.processing.ProcessingEnvironment;
import javax.lang.model.element.Modifier;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Elements;
import javax.lang.model.util.Types;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Objects;

public class UserDefinedTypeStatementSetterGenerator {

    private final Elements elements;
    private final Types types;
    private final ProcessingEnvironment processingEnv;

    public UserDefinedTypeStatementSetterGenerator(ProcessingEnvironment processingEnv) {
        this.processingEnv = processingEnv;
        this.elements = processingEnv.getElementUtils();
        this.types = processingEnv.getTypeUtils();
    }

    public void generate(TypeMirror typeMirror) {
        this.generateMapper(typeMirror);
        this.generateListMapper(typeMirror);
    }

    public void generateMapper(TypeMirror typeMirror) {
        var element = this.types.asElement(typeMirror);
        var packageName = this.elements.getPackageOf(element);
        var typeSpec = TypeSpec.classBuilder(CommonUtils.getOuterClassesAsPrefix(element) + element.getSimpleName().toString() + "_CassandraParameterColumnMapper")
            .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
            .addSuperinterface(ParameterizedTypeName.get(CassandraTypes.PARAMETER_COLUMN_MAPPER, TypeName.get(typeMirror)));
        var entity = Objects.requireNonNull(DbEntity.parseEntity(this.types, typeMirror));
        var constructor = MethodSpec.constructorBuilder().addModifiers(Modifier.PUBLIC);
        this.addMappers(typeSpec, constructor, entity);
        var apply = MethodSpec.methodBuilder("apply")
            .addModifiers(Modifier.PUBLIC)
            .addAnnotation(Override.class)
            .addParameter(CassandraTypes.SETTABLE_BY_NAME, "_stmt")
            .addParameter(int.class, "_index")
            .addParameter(TypeName.get(typeMirror), "_value");
        apply.beginControlFlow("if (_value == null)").addStatement("_stmt.setToNull(_index)").addStatement("return").endControlFlow();

        apply.addStatement("var _type = ($T) _stmt.getType(_index)", CassandraTypes.USER_DEFINED_TYPE);
        this.readIndexes(apply, entity);
        apply.addStatement("var _object = _type.newValue()");
        apply.addCode("\n");
        this.setFields(apply, entity);
        apply.addCode("\n");
        apply.addStatement("_stmt.setUdtValue(_index, _object)");

        typeSpec.addMethod(apply.build());
        typeSpec.addMethod(constructor.build());
        try {
            var javaFile = JavaFile.builder(packageName.getQualifiedName().toString(), typeSpec.build()).build();
            CommonUtils.safeWriteTo(this.processingEnv, javaFile);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    public void generateListMapper(TypeMirror typeMirror) {
        var element = this.types.asElement(typeMirror);
        var packageName = this.elements.getPackageOf(element);
        var listType = ParameterizedTypeName.get(CommonClassNames.list, TypeName.get(typeMirror));
        var typeSpec = TypeSpec.classBuilder(CommonUtils.getOuterClassesAsPrefix(element) + element.getSimpleName().toString() + "_List_CassandraParameterColumnMapper")
            .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
            .addSuperinterface(ParameterizedTypeName.get(CassandraTypes.PARAMETER_COLUMN_MAPPER, listType));
        var entity = Objects.requireNonNull(DbEntity.parseEntity(this.types, typeMirror));
        var constructor = MethodSpec.constructorBuilder().addModifiers(Modifier.PUBLIC);
        this.addMappers(typeSpec, constructor, entity);
        var apply = MethodSpec.methodBuilder("apply")
            .addModifiers(Modifier.PUBLIC)
            .addAnnotation(Override.class)
            .addParameter(CassandraTypes.SETTABLE_BY_NAME, "_stmt")
            .addParameter(int.class, "_index")
            .addParameter(listType, "_listValue");
        apply.beginControlFlow("if (_listValue == null)").addStatement("_stmt.setToNull(_index)").addStatement("return").endControlFlow();

        apply.addStatement("var _listType = ($T) _stmt.getType(_index)", CassandraTypes.LIST_TYPE);
        apply.addStatement("var _type = ($T) _listType.getElementType()", CassandraTypes.USER_DEFINED_TYPE);
        this.readIndexes(apply, entity);
        apply.addStatement("var _listResult = new $T<$T>(_listValue.size())", ArrayList.class, CassandraTypes.UDT_VALUE);
        apply.beginControlFlow("for (var _value : _listValue)");
        apply.addStatement("var _object = _type.newValue()");
        this.setFields(apply, entity);
        apply.addStatement("_listResult.add(_object)");
        apply.endControlFlow();
        apply.addCode("\n");
        apply.addStatement("_stmt.setList(_index, _listResult, $T.class)", CassandraTypes.UDT_VALUE);

        typeSpec.addMethod(apply.build());
        typeSpec.addMethod(constructor.build());
        try {
            var javaFile = JavaFile.builder(packageName.getQualifiedName().toString(), typeSpec.build()).build();
            CommonUtils.safeWriteTo(this.processingEnv, javaFile);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    private void readIndexes(MethodSpec.Builder apply, DbEntity entity) {
        for (var entityField : entity.entityFields()) {
            var fieldName = entityField.element().getSimpleName().toString();
            apply.addStatement("var $N = _type.firstIndexOf($S)", "_index_of_" + fieldName, entityField.columnName());
        }
    }

    private void setFields(MethodSpec.Builder apply, DbEntity entity) {
        for (var entityField : entity.entityFields()) {
            var fieldName = entityField.element().getSimpleName().toString();
            var index = CodeBlock.of("$N", "_index_of_" + fieldName);
            var nativeType = CassandraNativeTypes.findNativeType(TypeName.get(entityField.typeMirror()));
            if (nativeType != null) {
                if (entity.entityType() == DbEntity.EntityType.RECORD) {
                    apply.addStatement(nativeType.bind("_object", "_value." + fieldName + "()", index));
                } else {
                    apply.addStatement(nativeType.bind("_object", "_value.get" + CommonUtils.capitalize(fieldName) + "()", index));
                }
            } else {
                var mapperName = "_" + fieldName + "_mapper";
                if (entity.entityType() == DbEntity.EntityType.RECORD) {
                    apply.addStatement("this.$N.apply(_object, $L, _value.$N())", mapperName, index, fieldName);
                } else {
                    apply.addStatement("this.$N.apply(_object, $L, _value.get$N())", mapperName, index, CommonUtils.capitalize(fieldName));
                }
            }
        }
    }

    private void addMappers(TypeSpec.Builder typeSpec, MethodSpec.Builder constructor, DbEntity entity) {
        for (var entityField : entity.entityFields()) {
            var fieldName = entityField.element().getSimpleName().toString();
            var nativeType = CassandraNativeTypes.findNativeType(TypeName.get(entityField.typeMirror()));
            if (nativeType == null) {
                var mapperName = "_" + fieldName + "_mapper";
                // todo mapping annotation support?
                var mapperType = ParameterizedTypeName.get(CassandraTypes.PARAMETER_COLUMN_MAPPER, TypeName.get(entityField.typeMirror()));
                constructor.addParameter(mapperType, mapperName);
                constructor.addStatement("this.$N = $N", mapperName, mapperName);
                typeSpec.addField(mapperType, mapperName, Modifier.PRIVATE, Modifier.FINAL);
            }
        }
    }
}

File: database/database-symbol-processor/src/main/kotlin/ru/tinkoff/kora/database/symbol/processor/cassandra/udt/UserDefinedTypeStatementSetterGenerator.kt
package ru.tinkoff.kora.database.symbol.processor.cassandra.udt

import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.processing.SymbolProcessorEnvironment
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.ksp.toTypeName
import com.squareup.kotlinpoet.ksp.writeTo
import ru.tinkoff.kora.database.symbol.processor.cassandra.CassandraNativeTypes
import ru.tinkoff.kora.database.symbol.processor.cassandra.CassandraTypes
import ru.tinkoff.kora.database.symbol.processor.model.DbEntity
import ru.tinkoff.kora.ksp.common.KotlinPoetUtils.controlFlow
import ru.tinkoff.kora.ksp.common.generatedClassName

class UserDefinedTypeStatementSetterGenerator(private val environment: SymbolProcessorEnvironment) {

    fun generate(resolver: Resolver, classDeclaration: KSClassDeclaration) {
        this.generateSetter(resolver, classDeclaration)
        this.generateListSetter(resolver, classDeclaration)
    }

    private fun generateSetter(resolver: Resolver, classDeclaration: KSClassDeclaration) {
        val type = classDeclaration.asType(listOf())
        val typeName = type.toTypeName().copy(false)
        val entity = DbEntity.parseEntity(type)!!
        val typeSpec = TypeSpec.classBuilder(classDeclaration.generatedClassName("CassandraParameterColumnMapper"))
            .addModifiers(KModifier.PUBLIC, KModifier.FINAL)
            .addSuperinterface(CassandraTypes.parameterColumnMapper.parameterizedBy(typeName))
        val constructor = FunSpec.constructorBuilder()
        val apply = FunSpec.builder("apply")
            .addModifiers(KModifier.OVERRIDE)
            .addParameter("_stmt", CassandraTypes.settableByName)
            .addParameter("_index", Int::class.javaPrimitiveType!!.asTypeName())
            .addParameter("_value", typeName.copy(true))
        apply.controlFlow("if (_value == null)") {
            addStatement("_stmt.setToNull(_index)")
            addStatement("return")
        }
        apply.addStatement("val _type = _stmt.getType(_index) as %T", CassandraTypes.userDefinedType)
        apply.addStatement("val _object = _type.newValue()")
        apply.addCode("\n")
        addMappers(entity, typeSpec, constructor)
        getIndexes(entity, apply)
        setObject(entity, apply)
        apply.addStatement("_stmt.setUdtValue(_index, _object)")
        typeSpec.addFunction(apply.build())
        typeSpec.primaryConstructor(constructor.build())

        FileSpec.get(classDeclaration.packageName.asString(), typeSpec.build()).writeTo(environment.codeGenerator, false, listOfNotNull(classDeclaration.containingFile))
    }

    private fun generateListSetter(resolver: Resolver, classDeclaration: KSClassDeclaration) {
        val type = classDeclaration.asType(listOf())
        val typeName = type.toTypeName()
        val listTypeName = LIST.parameterizedBy(typeName.copy(false)).copy(false)
        val entity = DbEntity.parseEntity(type)!!
        val typeSpec = TypeSpec.classBuilder(classDeclaration.generatedClassName("List_CassandraParameterColumnMapper"))
            .addModifiers(KModifier.PUBLIC, KModifier.FINAL)
            .addSuperinterface(CassandraTypes.parameterColumnMapper.parameterizedBy(listTypeName))
        val constructor = FunSpec.constructorBuilder()
        val apply = FunSpec.builder("apply")
            .addModifiers(KModifier.OVERRIDE)
            .addParameter("_stmt", CassandraTypes.settableByName)
            .addParameter("_index", Int::class.javaPrimitiveType!!.asTypeName())
            .addParameter("_listValue", listTypeName.copy(true))
        apply.controlFlow("if (_listValue == null)") {
            addStatement("_stmt.setToNull(_index)")
            addStatement("return")
        }
        apply.addStatement("val _listType = _stmt.getType(_index) as %T", CassandraTypes.listType)
        apply.addStatement("val _type = _listType.getElementType() as %T", CassandraTypes.userDefinedType)
        getIndexes(entity, apply)
        addMappers(entity, typeSpec, constructor)
        typeSpec.primaryConstructor(constructor.build())
        apply.controlFlow("val _udtList = _listValue.map { _value ->") {
            apply.addStatement("val _object = _type.newValue()")
            apply.addCode("\n")
            setObject(entity, apply)
            addStatement("_object")
        }
        apply.addStatement("_stmt.setList(_index, _udtList, %T::class.java)", CassandraTypes.udtValue)
        typeSpec.addFunction(apply.build())

        FileSpec.get(classDeclaration.packageName.asString(), typeSpec.build()).writeTo(environment.codeGenerator, false, listOfNotNull(classDeclaration.containingFile))
    }

    private fun setObject(entity: DbEntity, apply: FunSpec.Builder) {
        for (field in entity.fields) {
            val fieldName = field.property.simpleName.asString()
            val fieldTypeName = field.type.toTypeName()
            apply.controlFlow("_value.%N.let {", fieldName) {
                if (field.type.isMarkedNullable) {
                    beginControlFlow("if (it == null)")
                    addStatement("_object.setToNull(%N)", "_index_of_$fieldName")
                    nextControlFlow("else")
                }
                val nativeType = CassandraNativeTypes.findNativeType(fieldTypeName)
                if (nativeType != null) {
                    addCode(nativeType.bind("_object", CodeBlock.of("it"), CodeBlock.of("%N", "_index_of_$fieldName")))
                    addCode("\n")
                } else {
                    val mapperName = "_${fieldName}_mapper"
                    addStatement("this.%N.apply(_object, %N, it)", mapperName, "_index_of_$fieldName")
                }
                if (field.type.isMarkedNullable) {
                    endControlFlow()
                }
            }
        }
    }

    fun getIndexes(entity: DbEntity, apply: FunSpec.Builder) {
        for (field in entity.columns) {
            val fieldName = field.property.simpleName.asString()
            apply.addStatement("val %N = _type.firstIndexOf(%S)", "_index_of_$fieldName", field.columnName)
        }
    }

    fun addMappers(entity: DbEntity, typeSpec: TypeSpec.Builder, constructor: FunSpec.Builder) {
        for (field in entity.fields) {
            val fieldName = field.property.simpleName.asString()
            val fieldTypeName = field.type.toTypeName()
            val nativeType = CassandraNativeTypes.findNativeType(fieldTypeName)
            if (nativeType == null) {
                val mapperName = "_${fieldName}_mapper"
                val mapperType = CassandraTypes.parameterColumnMapper.parameterizedBy(fieldTypeName.copy(false))
                constructor.addParameter(mapperName, mapperType)
                constructor.addStatement("this.%N = %N", mapperName, mapperName)

                typeSpec.addProperty(mapperName, mapperType, KModifier.PRIVATE, KModifier.FINAL)
            }
        }
    }
}
